### PR TITLE
compiler-rt: implement __udivei4 and __umodei4

### PR DESCRIPTION
**Pull request by jedisct1**
Allows dividing and formatting arbitrary-large unsigned integers.

The ability to print integers larger than `u128` is something I've been frequently in need of, if only for debugging purposes.

### CODE REVIEW

File: lib/compiler_rt/udivmodei4.zig [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    const u = u_p[0 .. bits / 32];
+    const v = v_p[0 .. bits / 32];
+    var r = r_p[0 .. bits / 32];
+    _ = divmod(null, r, u, v) catch unreachable;
+}
+
+test "__udivei4/__umodei4" {
+    var i: usize = 100000;
+    while (i > 0) : (i -= 1) {
+        const u = std.crypto.random.int(u1000);
```
Comment (matu3ba) [inline lib/compiler_rt/udivmodei4.zig:114]: Could we use a constant predefined seed here to keep tests reproducible? I guess you kept them running for a while?
crypto.random sounds also overkill. Why not
```
const RndGen = std.rand.DefaultPrng;
    var rnd = RndGen.init(42);
    var i: u32 = 0;
    while (i < 10_000) : (i += 1) {
        var rand_num = rnd.random().int(i64);
        try test__TESTNAME(rand_num);
    }
    ```
like [here](https://github.com/ziglang/zig/blob/0d66112643286a754101eb71382f860ae1874012/lib/compiler_rt/popcountdi2_test.zig#L28) ?

File: lib/compiler_rt/udivmodei4.zig [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+const shl = std.math.shl;
+
+const max_limbs = std.math.divCeil(usize, 65535, 32) catch unreachable; // max supported type is u65535
+
+comptime {
+    @export(__udivei4, .{ .name = "__udivei4", .linkage = common.linkage });
+    @export(__umodei4, .{ .name = "__umodei4", .linkage = common.linkage });
+}
+
+fn divmod(q: ?[]u32, r: ?[]u32, u: []const u32, v: []const u32) !void {
```
Comment (matu3ba) [inline lib/compiler_rt/udivmodei4.zig:13]: Might be nice to mention that this uses `Knuth's Algorithm D` for later review.
Please add `@setRuntimeSafe(is_test)` (https://ziglang.org/documentation/master/#toc-setRuntimeSafety).

File: lib/compiler_rt/udivmodei4.zig [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        r_[n] = shr(u32, un[n], s);
+    }
+}
+
+pub fn __udivei4(r_q: [*c]u32, u_p: [*c]const u32, v_p: [*c]const u32, bits: usize) callconv(.C) void {
+    const u = u_p[0 .. bits / 32];
+    const v = v_p[0 .. bits / 32];
+    var q = r_q[0 .. bits / 32];
+    _ = divmod(q, null, u, v) catch unreachable;
```
Comment (matu3ba) [inline lib/compiler_rt/udivmodei4.zig:101]: This should be a inline call.

File: lib/compiler_rt/udivmodei4.zig [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+pub fn __umodei4(r_p: [*c]u32, u_p: [*c]const u32, v_p: [*c]const u32, bits: usize) callconv(.C) void {
+    const u = u_p[0 .. bits / 32];
+    const v = v_p[0 .. bits / 32];
+    var r = r_p[0 .. bits / 32];
+    _ = divmod(null, r, u, v) catch unreachable;
```
Comment (matu3ba) [inline lib/compiler_rt/udivmodei4.zig:108]: This should be a inline call.

**Review by matu3ba — CHANGES_REQUESTED** [review]
Some general things to polish, but I didn't look up the algorithm d by knuth yet.

I guess the signed versions will follow after this is merged. See [here for llvm PR discussions](https://reviews.llvm.org/D120329).

File: lib/compiler_rt/udivmodei4.zig (hunk 1 above)
Comment (jedisct1) [reply lib/compiler_rt/udivmodei4.zig:114]: Randomized inputs help to catch cases that a deterministic sequence may not. This is exactly what happened here in CI.

File: lib/compiler_rt/udivmodei4.zig [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    if (r) |r_| {
+        i = 0;
+        while (i <= n) : (i += 1) {
+            r_[i] = shr(u32, un[i], s) | shl(u32, un[i + 1], 32 - s);
+        }
+        r_[n] = shr(u32, un[n], s);
+    }
+}
+
+pub fn __udivei4(r_q: [*c]u32, u_p: [*c]const u32, v_p: [*c]const u32, bits: usize) callconv(.C) void {
```
Comment (matu3ba) [inline lib/compiler_rt/udivmodei4.zig:97]: `@setRuntimeSafe(is_test)` with the inlined `divmod` function should be good enough to disable the bounds checks for all of it. Same applies for `__umodei4`.

File: lib/compiler_rt/udivmodei4.zig (hunk 1 above)
Comment (jedisct1) [reply lib/compiler_rt/udivmodei4.zig:114]: Changed to deterministic tests.

**Comment by jedisct1** [comment]
Thanks for your review, @matu3ba!

Yes, the signed versions will follow 😊

**Comment by andrewrk** [comment]
Tests failed for 32-bit MIPS. Perhaps an endianness issue (this target is big-endian)?

**Comment by jedisct1** [comment]
Ah, right. The limbs may be in big-endian order as well.

File: lib/compiler_rt/udivmodei4.zig [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    if (r) |r_| {
+        i = 0;
+        while (i <= n) : (i += 1) {
+            limb_set(r_, i, shr(u32, limb(&un, i), s) | shl(u32, limb(&un, i + 1), 32 - s));
+        }
+        limb_set(r_, n, shr(u32, limb(&un, n), s));
+    }
+}
+
+pub fn __udivei4(r_q: [*c]u32, u_p: [*c]const u32, v_p: [*c]const u32, bits: usize) callconv(.C) void {
```
Comment (andrewrk) [inline lib/compiler_rt/udivmodei4.zig:115]: Never use `[*c]` pointers.

**Pull request merged**